### ISSUE TITLE
C# Wrapper: Not compatible to .NET Framework 4.6.1

### ISSUE THREAD

**Issue by avolts** [google/or-tools#1221]
We want to integrate the SAT Solver in one of our projects (.NET Framework 4.6.1) and had several problems. At the moment it is not possible that we can do an upgrade to .NET Framework 4.7.1 which is compatible to .NET Standard, so we would run into unpredictable issues, if we want to include our Nuget package/ NetStandard project containing the SAT solver into this project.

 **Official Microsoft Statement:**

> “While NuGet considers .NET Framework 4.6.1 as supporting .NET Standard 1.5 through 2.0, there are several issues with consuming .NET Standard libraries that were built for those versions from .NET Framework 4.6.1 projects” (https://docs.microsoft.com/en-gb/dotnet/standard/net-standard)

We would like to make the C# Wrapper usable for multiple target frameworks and adjust the NuSpec files, so that it is possible to use the Wrapper not only for .NET Standard but also for .NET Framework 4.6.1.
Would you support a pull request for those changes?

**Comment by Mizux** [comment]
TL;DR .Net Framework is not officially supported by Microsoft on Unix platform (OS) and you must use the mono project  which is not up to date on all targeted distro (e.g. ubuntu 16.04 LTS), so since v6.8 IIRC we decided to migrate to the Microsoft backed dotnet Core SDK and the .Net Standard 2.0 platform.

You can see the process for generating a cross-platform (ed multiple RIDs) .Net package [here](https://github.com/google/or-tools/blob/stable/ortools/dotnet/README.md). If you are able to build a multi-platforms / multiple-rids only using the dotnet core sdk cli on all platforms, we may agree to integrate it but I don't think it will be possible...

**Comment by georghinkel** [comment]
> I mean if you use a <TargetFrameworks>netstandard2.0;net461</TargetFrameworks> in the .csproj files, you impose to all unix user to have mono installed. 

Why? This setting just means that the .NET library is compiled one more time. In particular, the build will output two _different_ .NET assemblies, one that targets .NET Standard and the other targeting .NET Framework 4.6.1, but sharing the same code. You would then add the .NET Framework 4.6.1 into `lib/net461` and the .NET Standard assembly to `lib/netstandard2.0`. Clients building against .NET Core on Unix won't see any difference, they only see the .NET Standard library as before.

**Comment by Mizux** [comment]
As you said the problem is not building against it (i.e consuming the package). Is building the package itself !!!

What I mean, is .Net Framework 4.6.1 is not supported by Microsoft on all unix platforms.
note: on Linux you would have this typical error (tested on my pet project https://github.com/Mizux/dotnet-native) where I used `<TargetFrameworks>netstandard2.0;net461</TargetFrameworks>`
```sh
./linux.sh
...
 Restore completed in 162.63 ms for /usr/local/google/home/corentinl/work/dotnet-native/src/runtime.linux-x64.Foo/runtime.linux-x64.Foo.csproj.
/usr/local/google/home/corentinl/dotnet/sdk/2.2.103/Microsoft.Common.CurrentVersion.targets(1179,5):
error MSB3644: The reference assemblies for framework ".NETFramework,Version=v4.6.1" were not found.
To resolve this, install the SDK or Targeting Pack for this framework version or retarget your application to a version of the framework for which you have the SDK or Targeting Pack installed.
Note that assemblies will be resolved from the Global Assembly Cache (GAC) and will be used in place of reference assemblies.
Therefore your assembly may not be correctly targeted for the framework you intend.
[/usr/local/google/home/corentinl/work/dotnet-native/src/runtime.linux-x64.Foo/runtime.linux-x64.Foo.csproj]
runtime.linux-x64.Foo -> /usr/local/google/home/corentinl/work/dotnet-native/src/runtime.linux-x64.Foo/bin/Debug/netstandard2.0/linux-x64/Mizux.Foo.dll
```

**Comment by georghinkel** [comment]
What do you think of adding `net461` as target framework only to the winx64 runtime project? This way, as a user you would have to reference the winx64 package directly instead of the meta-package, but that is probably OK if you target .NET Framework because a .NET Framework application only runs on Windows (Mono can still use the .NET Standard version).

**Comment by lperron** [comment]
Frankly, this is a big effort on our side.
We need to change all our build process (we use dotnet core). And it is not
future proof, we should concentrate on dotnet core 3.0 instead.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le jeu. 2 mai 2019 à 11:00, Georg Hinkel <notifications@github.com> a
écrit :

**Comment by georghinkel** [comment]
Hm, if you build the Windows rid package under Windows 10 v1511 or newer or Windows Server 2016 or newer, you definitely have .NET 4.6.1 installed as part of the OS. The .NET Core SDK does nothing more than forwarding the build request to MSBuild underneath, so I don't actually see why it would be an effort.

Do you use a different build system for packaging than for CI? Otherwise, I would be happy to create a PR and see whether that works with the build system.

**Comment by lperron** [comment]
Should be fixed with the last PR.
